README.md
# ruby-mysql

## Description

ruby-mysql is a MySQL client library.
It is written entirely in Ruby.
Therefore libmysqlclient is not required and no compilation is required during installation.

## Installation

```ruby
gem install ruby-mysql
```

## Synopsis

```ruby
require 'mysql'

my = Mysql.connect('mysql://username:password@hostname:port/dbname?charset=utf8mb4')
my.query("select col1, col2 from tblname").each do |col1, col2|
  p col1, col2
end
stmt = my.prepare('insert into tblname (col1,col2) values (?,?)')
stmt.execute 123, 'abc'
```

## Major incompatibility with 3.0

### Result values are now converted by default

| MySQL type          | Ruby class         |
|---------------------|--------------------|
| NULL                | NilClass           |
| INT                 | Integer            |
| DECIMAL             | BigDecimal         |
| FLOAT, DOUBLE       | Float              |
| DATE                | Date               |
| DATETIME, TIMESTAMP | Time               |
| TIME                | Float (as seconds) |
| YEAR                | Integer            |
| CHAR, VARCHAR       | String             |
| BIT                 | String             |
| TEXT, BLOB, JSON    | String             |

3.0:
```ruby
pp my.query('select 123,123.45,now(),cast(now() as date)').fetch.map{[_1, _1.class]}
#=> [["123", String],
#    ["123.45", String],
#    ["2022-11-15 00:17:11", String],
#    ["2022-11-15", String]]
```

4.0:
```ruby
pp my.query('select 123,123.45,now(),cast(now() as date)').fetch.map{[_1, _1.class]}
#=> [[123, Integer],
#    [0.12345e3, BigDecimal],
#    [2022-11-15 00:17:17 +0900, Time],
#    [#<Date: 2022-11-15 ((2459899j,0s,0n),+0s,2299161j)>, Date]]
```

To specify `cast: false`, you get the same behavior as in 3.0.
```ruby
my.query('select 123,123.45,now(),cast(now() as date)', cast: false).fetch.map{[_1, _1.class]}
#=> [["123", String],
#    ["123.45", String],
#    ["2022-11-15 00:19:18", String],
#    ["2022-11-15", String]]
```

It can also be specified during Mysql.new and Mysql.connect.

```ruby
my = Mysql.connect('mysql://user:pass@localhost/', cast: false)
```

Changing mysql.default_options will affect the behavior of subsequently created instances.

```ruby
my1 = Mysql.connect('mysql://user:pass@localhost/')
Mysql.default_options[:cast] = false
my2 = Mysql.connect('mysql://user:pass@localhost/')
pp my1.query('select 123,123.45,now(),cast(now() as date)').fetch.map{[_1, _1.class]}
#=> [[123, Integer],
#    [0.12345e3, BigDecimal],
#    [2022-11-15 00:26:09 +0900, Time],
#    [#<Date: 2022-11-15 ((2459899j,0s,0n),+0s,2299161j)>, Date]]
pp my2.query('select 123,123.45,now(),cast(now() as date)').fetch.map{[_1, _1.class]}
#=> [["123", String],
#    ["123.45", String],
#    ["2022-11-15 00:26:09", String],
#    ["2022-11-15", String]]
```

### Mysql::Result#each now always return records from the beginning

3.0:
```ruby
res = my.query('select 123 union select 456')
res.entries
#=> [["123"], ["456"]]
res.entries
#=> []
```

4.0:
```ruby
res = my.query('select 123 union select 456')
res.entries
#=> [[123], [456]]
res.entries
#=> [[123], [456]]
```

## Copyright

* Author: TOMITA Masahiro <tommy@tmtm.org>
* Copyright: Copyright 2008 TOMITA Masahiro
* License: MIT


CHANGELOG.md
## [4.1.0] - 2023-10-08

- Support geometry column <https://gitlab.com/tmtms/ruby-mysql/-/merge_requests/1>
- FIX: If the server is localhost and disconnect the connection, Errno::EPIPE is raised <https://gitlab.com/tmtms/ruby-mysql/-/merge_requests/3> <https://gitlab.com/tmtms/ruby-mysql/-/issues/1>
- Allow for existing socket when connecting <https://gitlab.com/tmtms/ruby-mysql/-/merge_requests/2> <https://gitlab.com/tmtms/ruby-mysql/-/merge_requests/5>
  - add `io` keyword parameter

## [4.0.0] - 2022-11-09

### Incompatible changes

- unsupport MySQL 5.5 and Ruby 2.5
- `Mysql::Result#fetch` returns converted values unless `cast` is false
- `Mysql::Result#each` always returns records from the beginning
- Retrieving a DECIMAL type with a prepared statement returns a `BigDecimal` object instead of `String`
- Retrieving a DATE type value with a prepared statement returns a `Date` object instead of `Time`
- delete `Mysql#more_results`. use `#more_results?` instead.
- remove `Mysql::Result#fetch_field`, `#field_tell`, `#field_seek`, `#fetch_field_direct`, `#fetch_lengths`, `#num_fields`.
- error 'command out of sync' is `Mysql::ClientError::CommandOutOfSync` instead of `RuntimeError`.
- error 'Authentication requires secure connection' is `Mysql::ClientError::AuthPluginErr` instead of `RuntimeError`.

### Features

- `Mysql.default_options` is global options.
- `Mysql#connect` option `ssl_mode`: support `SSL_MODE_VERIFY_CA`, `SSL_MODE_VERIFY_IDENTITY`.
- `Mysql#connect` option `ssl_context_params`: see `OpenSSL::SSL::SSLContext#set_params`.
- `Mysql#connect` option `connect_attrs`.
- `Mysql::Stmt#more_results?`, `#next_result`, `#info`.
- `Mysql#close` and `Mysql::Stmt#close` read pending packets.
- `Mysql#query` and `Mysql::Stmt#execute` option: `return_result` and `yield_null_result`.
- support session tracking. See https://dev.mysql.com/doc/refman/8.0/en/session-state-tracking.html
- thread safe.
- `Mysql#query`, `Mysql::Stmt#execute` option: `auto_store_result`.
- add `Mysql::Result#server_status`

### Fixes

- When using connection that disconnected from client. error 'MySQL client is not connected' is occured instead of 'MySQL server has gone away'.
- When SSL error, `MySQL::ClientError::ServerLost` or `ServerGoneError` is occured instead of `OpenSSL::SSL::SSLError`.
- `Mysql#server_version` don't require connection.
- use `connect_timeout` instead of `read/write_timeout` on initial negotiation.
- enable to changing `local_infile` for established connection.
- `Mysql.connect` with host nil ignores parameters
- raises `IOError` after `Mysql#close`
- Fractional seconds of time types were ignored when retrieving values using prepared statements
- `Mysql::Stmt#execute` allows true or false values.

### Others

- Mysql#prpare raises Mysql::ClientError if the connection is not connected.
- using rubocop
- migrate from GitHub to GitLab
- ignore Gemfile.lock
- use rspec instead of test-unit
- using connection parameter from spec/config.rb for testing
- split files

## [3.0.1] - 2022-06-18

- LICENSE: correct author
- FIX: correct LOAD DATA LOCAL INFILE result information.
- FIX: reset SERVER_MORE_RESULTS_EXISTS when error packet is received.
- FIX: close the socket when the connection is disconnected.
- FIX: allow multiple results by default.

## [3.0.0] - 2021-11-16

- `Mysql.new` no longer connect. use `Mysql.connect` or `Mysql#connect`.

- `Mysql.init` is removed. use `Mysql.new` instead.

- `Mysql.new`, `Mysql.conncet` and `Mysql#connect` takes URI object or URI string or Hash object.
  example:
      Mysql.connect('mysql://user:password@hostname:port/dbname?charset=ascii')
      Mysql.connect('mysql://user:password@%2Ftmp%2Fmysql.sock/dbname?charset=ascii') # for UNIX socket
      Mysql.connect('hostname', 'user', 'password', 'dbname')
      Mysql.connect(host: 'hostname', username: 'user', password: 'password', database: 'dbname')

- `Mysql.options` is removed. use `Mysql#param = value` instead.
  For example:
      m = Mysql.init
      m.options(Mysql::OPT_LOCAL_INFILE, true)
      m.connect(host, user, passwd)
  change to
      m = Mysql.new
      m.local_infile = true
      m.connect(host, user, passwd)
  or
      m = Mysql.connect(host, user, passwd, local_infile: true)

- `Mysql::Time` is removed.
  Instead, `Time` object is returned for the DATE, DATETIME, TIMESTAMP data,
  and `Integer` object is returned for the TIME data.
  If DATE, DATETIME, TIMESTAMP are invalid values for Time, nil is returned.

- meaningless methods are removed:
  * `bind_result`
  * `client_info`
  * `client_version`
  * `get_proto_info`
  * `get_server_info`
  * `get_server_version`
  * `proto_info`
  * `query_with_result`

- alias method are removed:
  * `get_host_info`: use `host_info`
  * `real_connect`: use `connect`
  * `real_query`: use `query`

- methods corresponding to deprecated APIs in MySQL are removed:
  * `list_dbs`: use `SHOW DATABASES`
  * `list_fields`: use `SHOW COLUMNS`
  * `list_processes`: use `SHOW PROCESSLIST`
  * `list_tables`: use `SHOW TABLES`


spec/spec_helper.rb
require 'rspec'
require 'power_assert'
require 'mysql'
require 'fileutils'

class RSpec::Core::ExampleGroup  # rubocop:disable Style/ClassAndModuleChildren
  def assert(&block)
    PowerAssert.start(block, assertion_method: __callee__) do |pa|
      result = pa.yield
      message = pa.message_proc.call
      unless result
        ex = RSpec::Expectations::ExpectationNotMetError.new(message)
        RSpec::Support.notify_failure(ex)
      end
    end
  end
end

conf = "#{__dir__}/config.rb"
File.write(conf, File.read("#{conf}.example")) unless File.exist? conf
load conf


spec/mysql/protocol_spec.rb
require 'spec_helper'

describe Mysql::Protocol do
  describe '.new' do
    before{ FileUtils.touch MYSQL_SOCKET }

    it 'uses the object provided as the `:io` option' do
      socket = Object.new
      protocol = described_class.new(host: 'localhost', io: socket)
      assert{ protocol.instance_variable_get(:@socket) == socket }
    end
  end
end


lib/mysql/stmt.rb
class Mysql
  # @!visibility public
  # Prepared statement
  # @!attribute [r] affected_rows
  #   @return [Integer]
  # @!attribute [r] insert_id
  #   @return [Integer]
  # @!attribute [r] server_status
  #   @return [Integer]
  # @!attribute [r] warning_count
  #   @return [Integer]
  # @!attribute [r] param_count
  #   @return [Integer]
  # @!attribute [r] fields
  #   @return [Array<Mysql::Field>]
  # @!attribute [r] sqlstate
  #   @return [String]
  class Stmt
    include Enumerable

    attr_reader :affected_rows, :info, :insert_id, :server_status, :warning_count
    attr_reader :param_count, :fields, :sqlstate

    # @private
    def self.finalizer(protocol, statement_id)
      proc do
        protocol.gc_stmt statement_id
      end
    end

    # @private
    # @param [Mysql::Protocol] protocol
    def initialize(protocol, **opts)
      @protocol = protocol
      @opts = opts
      @statement_id = nil
      @affected_rows = @insert_id = @server_status = @warning_count = 0
      @sqlstate = "00000"
      @param_count = nil
    end

    # @private
    # parse prepared-statement and return {Mysql::Stmt} object
    # @param [String] str query string
    # @return self
    def prepare(str)
      raise ClientError, 'MySQL client is not connected' unless @protocol
      close
      begin
        @sqlstate = "00000"
        @statement_id, @param_count, @fields = @protocol.stmt_prepare_command(str)
      rescue ServerError => e
        @last_error = e
        @sqlstate = e.sqlstate
        raise
      end
      ObjectSpace.define_finalizer(self, self.class.finalizer(@protocol, @statement_id))
      self
    end

    # Execute prepared statement.
    # @param [Object] values values passed to query
    # @return [Mysql::Result] if return_result is true and the query returns result set.
    # @return [nil] if return_result is true and the query does not return result set.
    # @return [self] if return_result is false or block is specified.
    def execute(*values, **opts, &block)
      raise ClientError, "Invalid statement handle" unless @statement_id
      raise ClientError, "not prepared" unless @param_count
      raise ClientError, "parameter count mismatch" if values.length != @param_count
      values = values.map{|v| @protocol.charset.convert v}
      opts = @opts.merge(opts)
      begin
        @sqlstate = "00000"
        @protocol.stmt_execute_command @statement_id, values
        @fields = @result = nil
        if block
          while true
            get_result
            res = store_result(**opts)
            block.call res if res || opts[:yield_null_result]
            break unless more_results?
          end
          return self
        end
        get_result
        return self unless opts[:return_result]
        return store_result(**opts)
      rescue ServerError => e
        @last_error = e
        @sqlstate = e.sqlstate
        raise
      end
    end

    def get_result
      @protocol.get_result
      @affected_rows, @insert_id, @server_status, @warning_count, @info =
        @protocol.affected_rows, @protocol.insert_id, @protocol.server_status, @protocol.warning_count, @protocol.message
    end

    def store_result(**opts)
      return nil if @protocol.field_count.nil? || @protocol.field_count == 0
      @fields = @protocol.retr_fields
      opts = @opts.merge(opts)
      @result = StatementResult.new(@fields, @protocol, **opts)
    end

    def more_results?
      @protocol.more_results?
    end

    # execute next query if precedure is called.
    # @return [Mysql::StatementResult] result set of query if return_result is true.
    # @return [true] if return_result is false and result exists.
    # @return [nil] query returns no results or no more results.
    def next_result(**opts)
      return nil unless more_results?
      opts = @opts.merge(opts)
      @fields = @result = nil
      get_result
      return self unless opts[:return_result]
      return store_result(**opts)
    rescue ServerError => e
      @last_error = e
      @sqlstate = e.sqlstate
      raise
    end

    # Close prepared statement
    # @return [void]
    def close
      ObjectSpace.undefine_finalizer(self)
      @protocol.stmt_close_command @statement_id if @statement_id
      @statement_id = nil
    end

    # @return [Array] current record data
    def fetch(**opts)
      @result.fetch(**opts)
    end

    # Return data of current record as Hash.
    # The hash key is field name.
    # @param [Boolean] with_table if true, hash key is "table_name.field_name".
    # @return [Hash] record data
    def fetch_hash(**opts)
      @result.fetch_hash(**opts)
    end

    # Iterate block with record.
    # @yield [Array] record data
    # @return [Mysql::Stmt] self
    # @return [Enumerator] If block is not specified
    def each(**opts, &block)
      return enum_for(:each, **opts) unless block
      while (rec = fetch(*opts))
        block.call rec
      end
      self
    end

    # Iterate block with record as Hash.
    # @param [Boolean] with_table if true, hash key is "table_name.field_name".
    # @yield [Hash] record data
    # @return [Mysql::Stmt] self
    # @return [Enumerator] If block is not specified
    def each_hash(**opts, &block)
      return enum_for(:each_hash, **opts) unless block
      while (rec = fetch_hash(**opts))
        block.call rec
      end
      self
    end

    # @return [Integer] number of record
    def size
      @result.size
    end
    alias num_rows size

    # Set record position
    # @param [Integer] n record index
    # @return [void]
    def data_seek(n)
      @result.data_seek(n)
    end

    # @return [Integer] current record position
    def row_tell
      @result.row_tell
    end

    # Set current position of record
    # @param [Integer] n record index
    # @return [Integer] previous position
    def row_seek(n)
      @result.row_seek(n)
    end

    # @return [Integer] number of columns for last query
    def field_count
      @fields.length
    end

    # ignore
    # @return [void]
    def free_result
      # dummy
    end

    # Returns Mysql::Result object that is empty.
    # Use fields to get list of fields.
    # @return [Mysql::Result]
    def result_metadata
      return nil if @fields.empty?
      Result.new @fields
    end
  end
end


lib/mysql/field.rb
class Mysql
  # @!visibility public
  # Field class
  class Field
    # @return [String] database name
    attr_reader :db
    # @return [String] table name
    attr_reader :table
    # @return [String] original table name
    attr_reader :org_table
    # @return [String] field name
    attr_reader :name
    # @return [String] original field name
    attr_reader :org_name
    # @return [Integer] charset id number
    attr_reader :charsetnr
    # @return [Integer] field length
    attr_reader :length
    # @return [Integer] field type
    attr_reader :type
    # @return [Integer] flag
    attr_reader :flags
    # @return [Integer] number of decimals
    attr_reader :decimals
    # @return [String] defualt value
    attr_reader :default
    alias def default

    # @private
    attr_accessor :result

    # @attr [Protocol::FieldPacket] packet
    def initialize(packet)
      @db, @table, @org_table, @name, @org_name, @charsetnr, @length, @type, @flags, @decimals, @default =
        packet.db, packet.table, packet.org_table, packet.name, packet.org_name, packet.charsetnr, packet.length, packet.type, packet.flags, packet.decimals, packet.default
      @db.force_encoding('utf-8')
      @table.force_encoding('utf-8')
      @org_table.force_encoding('utf-8')
      @name.force_encoding('utf-8')
      @org_name.force_encoding('utf-8')
      @flags |= NUM_FLAG if is_num_type?
      @max_length = nil
    end

    # @return [Hash] field information
    def to_hash
      {
        "name"       => @name,
        "table"      => @table,
        "def"        => @default,
        "type"       => @type,
        "length"     => @length,
        "max_length" => max_length,
        "flags"      => @flags,
        "decimals"   => @decimals,
      }
    end

    # @private
    def inspect
      "#<Mysql::Field:#{@name}>"
    end

    # @return [Boolean] true if numeric field.
    def is_num?
      @flags & NUM_FLAG != 0
    end

    # @return [Boolean] true if not null field.
    def is_not_null?
      @flags & NOT_NULL_FLAG != 0
    end

    # @return [Boolean] true if primary key field.
    def is_pri_key?
      @flags & PRI_KEY_FLAG != 0
    end

    # @return [Integer] maximum width of the field for the result set
    def max_length
      return @max_length if @max_length
      @max_length = 0
      @result&.calculate_field_max_length
      @max_length
    end

    attr_writer :max_length

    private

    def is_num_type?
      [TYPE_DECIMAL, TYPE_TINY, TYPE_SHORT, TYPE_LONG, TYPE_FLOAT, TYPE_DOUBLE, TYPE_LONGLONG, TYPE_INT24].include?(@type) || (@type == TYPE_TIMESTAMP && (@length == 14 || @length == 8))
    end
  end
end


lib/mysql.rb
# coding: ascii-8bit

# Copyright (C) 2008 TOMITA Masahiro
# mailto:tommy@tmtm.org

require 'uri'

# MySQL connection class.
# @example
#  my = Mysql.connect('hostname', 'user', 'password', 'dbname')
#  res = my.query 'select col1,col2 from tbl where id=123'
#  res.each do |c1, c2|
#    p c1, c2
#  end
class Mysql
  require_relative "mysql/field"
  require_relative "mysql/result"
  require_relative "mysql/stmt"
  require_relative "mysql/constants"
  require_relative "mysql/error"
  require_relative "mysql/charset"
  require_relative "mysql/protocol"
  require_relative "mysql/packet"

  VERSION            = -'4.1.0'             # Version number of this library
  MYSQL_UNIX_PORT    = -"/tmp/mysql.sock"   # UNIX domain socket filename
  MYSQL_TCP_PORT     = 3306                # TCP socket port number

  # @!attribute [rw] host
  #   @return [String, nil]
  # @!attribute [rw] username
  #   @return [String, nil]
  # @!attribute [rw] password
  #   @return [String, nil]
  # @!attribute [rw] database
  #   @return [String, nil]
  # @!attribute [rw] port
  #   @return [Integer, String, nil]
  # @!attribute [rw] socket
  #   @return [String, nil] socket filename
  # @!attribute [rw] flags
  #   @return [Integer, nil]
  # @!attribute [rw] io
  #   @return [[BasicSocket, OpenSSL::SSL::SSLSocket], nil]
  # @!attribute [rw] connect_timeout
  #   @return [Numeric, nil]
  # @!attribute [rw] read_timeout
  #   @return [Numeric, nil]
  # @!attribute [rw] write_timeout
  #   @return [Numeric, nil]
  # @!attribute [rw] init_command
  #   @return [String, nil]
  # @!attribute [rw] local_infile
  #   @return [Boolean]
  # @!attribute [rw] load_data_local_dir
  #   @return [String, nil]
  # @!attribute [rw] ssl_mode
  #   @return [String, Integer] 1 or "disabled" / 2 or "preferred" / 3 or "required"
  # @!attribute [rw] ssl_context_params
  #   @return [Hash] See OpenSSL::SSL::Context#set_params
  # @!attribute [rw] get_server_public_key
  #   @return [Boolean]
  # @!attribute [rw] connect_attrs
  #   @return [Hash]
  # @!attribute [rw] yield_null_result
  #   @return [Boolean]
  # @!attribute [rw] return_result
  #   @return [Boolean]
  # @!attribute [rw] with_table
  #   @return [Boolean]
  # @!attribute [rw] auto_store_result
  #   @return [Boolean]
  # @!attribute [rw] cast
  #   @return [Boolean]
  DEFAULT_OPTS = {
    host: nil,
    username: nil,
    password: nil,
    database: nil,
    port: nil,
    socket: nil,
    flags: 0,
    io: nil,
    charset: nil,
    connect_timeout: nil,
    read_timeout: nil,
    write_timeout: nil,
    init_command: nil,
    local_infile: nil,
    load_data_local_dir: nil,
    ssl_mode: SSL_MODE_PREFERRED,
    ssl_context_params: {},
    get_server_public_key: false,
    connect_attrs: {},
    yield_null_result: true,
    return_result: true,
    with_table: false,
    auto_store_result: true,
    cast: true,
  }.freeze

  # @private
  attr_reader :protocol

  # @return [Array<Mysql::Field>] fields of result set
  attr_reader :fields

  # @return [Mysql::Result]
  attr_reader :result

  class << self
    # Make Mysql object and connect to mysqld.
    # parameter is same as arguments for {#initialize}.
    # @return [Mysql]
    def connect(*args, **opts)
      self.new(*args, **opts).connect
    end

    # Escape special character in string.
    # @param [String] str
    # @return [String]
    def escape_string(str)
      str.gsub(/[\0\n\r\\'"\x1a]/) do |s|
        case s
        when "\0" then "\\0"
        when "\n" then "\\n"
        when "\r" then "\\r"
        when "\x1a" then "\\Z"
        else "\\#{s}"
        end
      end
    end
    alias quote escape_string

    def default_options
      @default_options ||= DEFAULT_OPTS.dup
    end
  end

  # @overload initialize(uri, **opts)
  #   @param uri [String, URI] "mysql://username:password@host:port/database?param=value&..." / "mysql://username:password@%2Ftmp%2Fmysql.sock/database" / "mysql://username:password@/database?socket=/tmp/mysql.sock"
  #   @param opts [Hash] options
  # @overload initialize(host, username, password, database, port, socket, flags, **opts)
  #   @param host [String] hostname mysqld running
  #   @param username [String] username to connect to mysqld
  #   @param password [String] password to connect to mysqld
  #   @param database [String] initial database name
  #   @param port [String] port number (used if host is not 'localhost' or nil)
  #   @param socket [String] socket filename (used if host is 'localhost' or nil)
  #   @param flags [Integer] connection flag. Mysql::CLIENT_* ORed
  #   @param opts [Hash] options
  # @overload initialize(host: nil, username: nil, password: nil, database: nil, port: nil, socket: nil, flags: nil, **opts)
  #   @param host [String] hostname mysqld running
  #   @param username [String] username to connect to mysqld
  #   @param password [String] password to connect to mysqld
  #   @param database [String] initial database name
  #   @param port [String] port number (used if host is not 'localhost' or nil)
  #   @param socket [String] socket filename (used if host is 'localhost' or nil)
  #   @param flags [Integer] connection flag. Mysql::CLIENT_* ORed
  #   @param opts [Hash] options
  #   @option opts :host [String] hostname mysqld running
  #   @option opts :username [String] username to connect to mysqld
  #   @option opts :password [String] password to connect to mysqld
  #   @option opts :database [String] initial database name
  #   @option opts :port [String] port number (used if host is not 'localhost' or nil)
  #   @option opts :socket [String] socket filename (used if host is 'localhost' or nil)
  #   @option opts :flags [Integer] connection flag. Mysql::CLIENT_* ORed
  #   @option opts :charset [Mysql::Charset, String] character set
  #   @option opts :connect_timeout [Numeric, nil]
  #   @option opts :read_timeout [Numeric, nil]
  #   @option opts :write_timeout [Numeric, nil]
  #   @option opts :local_infile [Boolean]
  #   @option opts :load_data_local_dir [String]
  #   @option opts :ssl_mode [Integer]
  #   @option opts :ssl_context_params [Hash<Symbol, String>]
  #   @option opts :get_server_public_key [Boolean]
  #   @option opts :connect_attrs [Hash]
  #   @option opts :io [BasicSocket, OpenSSL::SSL::SSLSocket] Existing socket instance that will be used instead of creating a new socket
  def initialize(*args, **opts)
    @fields = nil
    @result = nil
    @protocol = nil
    @sqlstate = "00000"
    @host_info = nil
    @last_error = nil
    @opts = Mysql.default_options.dup
    parse_args(args, opts)
  end

  # Connect to mysqld.
  # parameter is same as arguments for {#initialize}.
  # @return [Mysql] self
  def connect(*args, **opts)
    parse_args(args, opts)
    if @opts[:flags] & CLIENT_COMPRESS != 0
      warn 'unsupported flag: CLIENT_COMPRESS' if $VERBOSE
      @opts[:flags] &= ~CLIENT_COMPRESS
    end
    @protocol = Protocol.new(@opts)
    @protocol.authenticate
    @host_info = (@opts[:host].nil? || @opts[:host] == "localhost") ? 'Localhost via UNIX socket' : "#{@opts[:host]} via TCP/IP"
    query @opts[:init_command] if @opts[:init_command]
    return self
  end

  def parse_args(args, opts)
    unless args.empty?
      case args[0]
      when URI
        uri = args[0]
      when /\Amysql:\/\//
        uri = URI.parse(args[0])
      when String, nil
        @opts[:host], user, passwd, dbname, port, socket, flags = *args
        @opts[:username] = user if user
        @opts[:password] = passwd if passwd
        @opts[:database] = dbname if dbname
        @opts[:port] = port if port
        @opts[:socket] = socket if socket
        @opts[:flags] = flags if flags
      when Hash
        # skip
      end
    end
    if uri
      host = uri.hostname.to_s
      host = URI.decode_www_form_component(host)
      if host.start_with?('/')
        @opts[:socket] = host
        host = ''
      end
      @opts[:host] = host
      @opts[:username] = URI.decode_www_form_component(uri.user.to_s)
      @opts[:password] = URI.decode_www_form_component(uri.password.to_s)
      @opts[:database] = uri.path.sub(/\A\/+/, '')
      @opts[:port] = uri.port
      opts = URI.decode_www_form(uri.query).to_h.transform_keys(&:intern).merge(opts) if uri.query
      opts[:flags] = opts[:flags].to_i if opts[:flags]
    end
    if args.last.kind_of? Hash
      opts = opts.merge(args.last)
    end
    @opts.update(opts)
  end

  DEFAULT_OPTS.each_key do |var|
    next if var == :charset
    define_method(var){@opts[var]}
    define_method("#{var}="){|val| @opts[var] = val}
  end

  # Disconnect from mysql.
  # @return [Mysql] self
  def close
    if @protocol
      @protocol.quit_command
      @protocol = nil
    end
    return self
  end

  # Disconnect from mysql without QUIT packet.
  # @return [Mysql] self
  def close!
    if @protocol
      @protocol.close
      @protocol = nil
    end
    return self
  end

  # Escape special character in MySQL.
  #
  # @param [String] str
  # return [String]
  def escape_string(str)
    self.class.escape_string str
  end
  alias quote escape_string

  # @return [Mysql::Charset] character set of MySQL connection
  def charset
    @opts[:charset]
  end

  # Set charset of MySQL connection.
  # @param [String, Mysql::Charset] cs
  def charset=(cs)
    charset = cs.is_a?(Charset) ? cs : Charset.by_name(cs)
    if @protocol
      @protocol.charset = charset
      query "SET NAMES #{charset.name}"
    end
    @opts[:charset] = charset
  end

  # @return [String] charset name
  def character_set_name
    @protocol.charset.name
  end

  # @return [Integer] last error number
  def errno
    @last_error ? @last_error.errno : 0
  end

  # @return [String] last error message
  def error
    @last_error&.error
  end

  # @return [String] sqlstate for last error
  def sqlstate
    @last_error ? @last_error.sqlstate : "00000"
  end

  # @return [Integer] number of columns for last query
  def field_count
    @fields.size
  end

  # @return [String] connection type
  def host_info
    @host_info
  end

  # @return [String] server version
  def server_info
    check_connection
    @protocol.server_info
  end

  # @return [Integer] server version
  def server_version
    @protocol&.server_version
  end

  # @return [String] information for last query
  def info
    @protocol&.message
  end

  # @return [Integer] number of affected records by insert/update/delete.
  def affected_rows
    @protocol ? @protocol.affected_rows : 0
  end

  # @return [Integer] latest auto_increment value
  def insert_id
    @protocol ? @protocol.insert_id : 0
  end

  # @return [Integer] number of warnings for previous query
  def warning_count
    @protocol ? @protocol.warning_count : 0
  end

  # Kill query.
  # @param [Integer] pid thread id
  # @return [Mysql] self
  def kill(pid)
    check_connection
    @protocol.kill_command pid
    self
  end

  # Execute query string.
  # @param str [String] Query.
  # @param return_result [Boolean]
  # @param yield_null_result [Boolean]
  # @return [Mysql::Result] if return_result is true and the query returns result set.
  # @return [nil] if return_result is true and the query does not return result set.
  # @return [self] if return_result is false or block is specified.
  # @example
  #  my.query("select 1,NULL,'abc'").fetch  # => [1, nil, "abc"]
  #  my.query("select 1,NULL,'abc'"){|res| res.fetch}
  def query(str, **opts, &block)
    opts = @opts.merge(opts)
    check_connection
    @fields = nil
    begin
      @protocol.query_command str
      if block
        while true
          @protocol.get_result
          res = store_result(**opts)
          block.call res if res || opts[:yield_null_result]
          break unless more_results?
        end
        return self
      end
      @protocol.get_result
      return self unless opts[:return_result]
      return store_result(**opts)
    rescue ServerError => e
      @last_error = e
      @sqlstate = e.sqlstate
      raise
    end
  end

  # Get all data for last query.
  # @return [Mysql::Result]
  # @return [nil] if no results
  def store_result(**opts)
    return nil if @protocol.field_count.nil? || @protocol.field_count == 0
    @fields = @protocol.retr_fields
    opts = @opts.merge(opts)
    @result = Result.new(@fields, @protocol, **opts)
  end

  # @return [Integer] Thread ID
  def thread_id
    check_connection
    @protocol.thread_id
  end

  # Set server option.
  # @param [Integer] opt {Mysql::OPTION_MULTI_STATEMENTS_ON} or {Mysql::OPTION_MULTI_STATEMENTS_OFF}
  # @return [Mysql] self
  def set_server_option(opt)
    check_connection
    @protocol.set_option_command opt
    self
  end

  # @return [Boolean] true if multiple queries are specified and unexecuted queries exists.
  def more_results?
    @protocol.more_results?
  end

  # execute next query if multiple queries are specified.
  # @return [Mysql::Result] result set of query if return_result is true.
  # @return [true] if return_result is false and result exists.
  # @return [nil] query returns no results.
  def next_result(**opts)
    return nil unless more_results?
    opts = @opts.merge(opts)
    @protocol.get_result
    @fields = nil
    return store_result(**opts) if opts[:return_result]
    true
  end

  # Parse prepared-statement.
  # @param [String] str query string
  # @return [Mysql::Stmt] Prepared-statement object
  def prepare(str, **opts)
    opts = @opts.merge(opts)
    st = Stmt.new(@protocol, **opts)
    st.prepare str
    st
  end

  # @private
  # Make empty prepared-statement object.
  # @return [Mysql::Stmt] If block is not specified.
  def stmt(**opts)
    opts = @opts.merge(opts)
    Stmt.new(@protocol, **opts)
  end

  # Check whether the  connection is available.
  # @return [Mysql] self
  def ping
    check_connection
    @protocol.ping_command
    self
  end

  # Flush tables or caches.
  # @param [Integer] op operation. Use Mysql::REFRESH_* value.
  # @return [Mysql] self
  def refresh(op)
    check_connection
    @protocol.refresh_command op
    self
  end

  # Reload grant tables.
  # @return [Mysql] self
  def reload
    refresh Mysql::REFRESH_GRANT
  end

  # Select default database
  # @return [Mysql] self
  def select_db(db)
    query "use #{db}"
    self
  end

  # shutdown server.
  # @return [Mysql] self
  def shutdown(level=0)
    check_connection
    @protocol.shutdown_command level
    self
  end

  # @return [String] statistics message
  def stat
    @protocol ? @protocol.statistics_command : 'MySQL server has gone away'
  end

  # Commit transaction
  # @return [Mysql] self
  def commit
    query 'commit'
    self
  end

  # Rollback transaction
  # @return [Mysql] self
  def rollback
    query 'rollback'
    self
  end

  # Set autocommit mode
  # @param [Boolean] flag
  # @return [Mysql] self
  def autocommit(flag)
    query "set autocommit=#{flag ? 1 : 0}"
    self
  end

  # session track
  # @return [Hash]
  def session_track
    @protocol.session_track
  end

  private

  def check_connection
    raise ClientError, 'MySQL client is not connected' unless @protocol
  end
end
